### PR TITLE
Decouple testing framework API from Java test specific frameworks

### PR DESCRIPTION
**Pull request by bartlomiej-laczkowski**
### What does this PR do?

This PR is a proposal for decoupling testing framework core API from Java specific testing frameworks.

### What issues does this PR fix or reference?

[#3452](https://github.com/eclipse/che/issues/3452)

### Previous behavior

In current state all of the testing frameworks that are Java specific are tightly coupled with the testing framework core API, which complicates "straight-forward" adding of new testing frameworks (i.e. PHPUnit). 

### New behavior

- Testing core framework API was moved to Che Core API packages (I was using the debugger API analogy)
- All code in testing framework core API that was tightly coupled to Java specific testing frameworks (i.e. JUnit, TestNG) was unbound and moved to related plug-ins 

That's more or less the general description of the changes that I have made. I am eager to answer all of the possible questions related to the details.

Signed-off-by: Bartlomiej Laczkowski bartlomiej.l@zend.com

### Changelog:

Decoupling of testing framework core API from Java specific testing frameworks implementation.

### Release Notes:

N/A refactoring of existing code.

### Docs PR:

None : same behavior than before

### CODE REVIEW

**Comment by skabashnyuk** [comment]
ci-test

**Comment by skabashnyuk** [comment]
ci-build

File: plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/src/main/java/org/eclipse/che/plugin/testing/junit/server/JUnitTestRunner.java [hunk 1]
```diff
@@ -49,29 +50,84 @@
 
     private static final String JUNIT4X_RUNNER_CLASS = "org.junit.runner.JUnitCore";
     private static final String JUNIT3X_RUNNER_CLASS = "junit.textui.TestRunner";
-    private String projectPath;
+    private String projectAbsolutePath;
```
Comment (vparfonov) [inline plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/src/main/java/org/eclipse/che/plugin/testing/junit/server/JUnitTestRunner.java]: can we avoid using project path as field?

File: plugins/plugin-testing-java/plugin-testing-testng/che-plugin-testing-testng-server/src/main/java/org/eclipse/che/plugin/testing/testng/server/TestNGRunner.java [hunk 2]
```diff
@@ -47,13 +49,59 @@
...
  *
  * </pre>
+ * 
  * @author Mirage Abeysekara
  */
 public class TestNGRunner implements TestRunner {
 
-    String projectPath;
-    ClassLoader projectClassLoader;
+    private String absoluteProjectPath;
```
Comment (vparfonov) [inline plugins/plugin-testing-java/plugin-testing-testng/che-plugin-testing-testng-server/src/main/java/org/eclipse/che/plugin/testing/testng/server/TestNGRunner.java]: can we avoid use project path as field? Maybe better put it as parameters for methods

**Review by vparfonov — COMMENTED** [review]
ci-build

File: plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/src/main/java/org/eclipse/che/plugin/testing/junit/server/JUnitTestRunner.java (hunk 1 above)
Comment (bartlomiej-laczkowski) [reply plugins/plugin-testing-java/plugin-testing-junit/che-plugin-testing-junit-server/src/main/java/org/eclipse/che/plugin/testing/junit/server/JUnitTestRunner.java]: Sure, no problem.

File: plugins/plugin-testing-java/plugin-testing-testng/che-plugin-testing-testng-server/src/main/java/org/eclipse/che/plugin/testing/testng/server/TestNGRunner.java (hunk 2 above)
Comment (bartlomiej-laczkowski) [reply plugins/plugin-testing-java/plugin-testing-testng/che-plugin-testing-testng-server/src/main/java/org/eclipse/che/plugin/testing/testng/server/TestNGRunner.java]: Sure, no problem.

**Comment by bartlomiej-laczkowski** [comment]
@vparfonov Hello, is there any chance that someone will verify this patch anytime soon? I would like to move on with adding support for PHP unit testing but without this change I can not move forward...

**Comment by TylerJewell** [comment]
@bartlomiej-laczkowski - sorry about this.  Vitalli's team has been swamped with two significant projects related to intelligent commands and electron client packaging. We'll ask someone to take this soon and get it merged.

Last week we changed our maintainer process so that we ask all PRs to include entries for the Changelog, Release Notes, and PR for the docs, if applicable.  Can you update your PR description with these items?

I'll be nominating you for committer status once this PR is merged. Our policy is that two major PRs gets you committer status.

**Comment by bartlomiej-laczkowski** [comment]
@TylerJewell - Thanks for the info! I have updated the PR description and will wait for this PR to be merged.

**Comment by vparfonov** [comment]
@bartlomiej-laczkowski thanks for your PR. Basicly look ok for me.
Your PR will be merged on end of this week after QA checking on our side

**Comment by vparfonov** [comment]
will be merged in https://github.com/eclipse/che/pull/3999

**Pull request closed without merging**